src/easydynamics/utils/convolution.py

```diff
@@ -0,0 +1,597 @@
+            x_unit=x_unit,
+            normalize_detailed_balance=normalize_detailed_balance,
+        )
+    else:
+        raise ValueError(
+            f"Unknown convolution method: {method}. Choose from 'analytical', or 'numerical'."
+        )
+
+
+def _numerical_convolution(
+    x: np.ndarray,
+    sample_model: Union[SampleModel, ModelComponent],
+    resolution_model: Union[SampleModel, ModelComponent],
+    offset: Optional[float] = 0.0,
+    upsample_factor: Optional[int] = 5,
+    extension_factor: Optional[float] = 0.2,
+    temperature: Optional[Union[Parameter, float]] = None,
+    temperature_unit: Optional[Union[str, sc.Unit]] = "K",
+    x_unit: Optional[Union[str, sc.Unit]] = "meV",
+    normalize_detailed_balance: Optional[bool] = True,
+) -> np.ndarray:
+    """
+    Numerical convolution using FFT with optional upsampling + extended range.
+    Includes detailed balance correction if temperature is provided.
+
+    Args:
+        x : np.ndarray
+            1D array of x values where the convolution is evaluated.
+        sample_model : SampleModel or ModelComponent
+            The sample model to be convolved.
+        resolution_model : SampleModel or ModelComponent
+            The resolution model to convolve with.
+        offset : Parameter, float, or None, optional
+            The offset to apply to the input array.
+        upsample_factor : int, optional
+            The factor by which to upsample the input data before convolution. Default is 5.
+        extension_factor : float, optional
+            The factor by which to extend the input data range before convolution. Default is 0.2.
+        temperature : Parameter, float, or None, optional
+            The temperature to use for detailed balance correction. Default is None.
+        temperature_unit : str or sc.Unit, optional
+            The unit of the temperature parameter. Default is 'K'.
+        x_unit : str or sc.Unit, optional
+            The unit of the x parameter. Default is 'meV'.
+        normalize_detailed_balance : bool, optional
+            Whether to normalize the detailed balance factor. Default is True.
+    Returns:
+        np.ndarray
+            The convolved values evaluated at x.
+    """
+
+    # Build dense grid
+    if upsample_factor == 0:
+        # Check if the array is uniformly spaced.
+        x_diff = np.diff(x)
+        is_uniform = np.allclose(x_diff, x_diff[0])
+        if not is_uniform:
+            raise ValueError(
+                "Input array `x` must be uniformly spaced if upsample_factor = 0."
+            )
+        x_dense = x
+    else:
+        # Create an extended and upsampled x grid
+        x_min, x_max = x.min(), x.max()
+        span = x_max - x_min
+        extra = extension_factor * span
+        extended_min = x_min - extra
+        extended_max = x_max + extra
+        num_points = len(x) * upsample_factor
+        x_dense = np.linspace(extended_min, extended_max, num_points)
```

Comment:
This is a pretty complex grid routine. Please consider extracting it to a helper function, where it can be properly unit tested.